### ISSUE TITLE
awaiting ShowMetroDialogAsync  has no effect

### ISSUE THREAD

**Issue by IsNull** [MahApps/MahApps.Metro#1910]
_Using the latest nuget package of mahapps (1.1.2.0), .NET 4.5_

I wonder why custom dialoges behave differently than the metro message box, in that the custom dialoges Task does not wait for the dialoge to close. I have found similar questions & bug reports but none of them have a acceptable answer, such as #1227.

```csharp
var controller = await _metroWindow.ShowMessageAsync( ... );
// As expected, awaits until the user has choosen a button
```
```csharp
await _metroWindow.ShowMetroDialogAsync( ... );
// await has no effect at all - behaves like a fire and forget async call
```

* Is this by design? Then why is there a task returned on ShowMetroDialogAsync ?
* Is this a bug? I may be able to look into this and do a pull request.
* Or is there another way to await the user response from this dialog asynchronously.

Can someone please share some thoughts behind this API?

**Comment by flagbug** [comment]
Yes, this is by the design, it's noted in the method summary: https://github.com/MahApps/MahApps.Metro/blob/master/MahApps.Metro/Controls/Dialogs/DialogManager.cs#L292

**Comment by IsNull** [comment]
If I see that correctly in the code, you did not implement awaiting since you had the problem that "dialog.WaitForButtonPressAsync()" is not available in the custom dialog?

Maybe it can be solved generically by hooking into the closing of the dialog, i.e. instead of waiting for a button pressed, wait until the dialog is gone...

I am happy to give it a shot and implement this bit, so the returned task actually awaits the closing of the dialog. Would you mind a pull request? Or do you have other plans with this api?

**Comment by IsNull** [comment]
After some testing, I finally have it working now in my fork https://github.com/ElderByte-/MahApps.Metro/commit/62557596a266137b5d19b067afd61560a110dd89

I added a simple method to the BaseMetroDialog
```csharp
        public Task WaitUnitlClosed()
        {
            TaskCompletionSource<object> tcs = new TaskCompletionSource<object>();

            Unloaded += (s,e) =>
            {
                tcs.TrySetResult(null);
            };

            return tcs.Task;
        }
```

With this, I added a new Method to the DialogManager (optionally we can also replace the old Async Dialog Method)

```csharp
        /// <summary>
        /// Adds a Metro Dialog instance to the specified window and makes it visible.
        /// You can await until this dialog is removed/hidden by awaiting the returned task.
        /// <para>You have to close the resulting dialog yourself with <see cref="HideMetroDialogAsync"/>.</para>
        /// </summary>
        /// <param name="window">The owning window of the dialog.</param>
        /// <param name="dialog">The dialog instance itself.</param>
        /// <param name="settings">An optional pre-defined settings instance.</param>
        /// <returns>A task representing the operation.</returns>
        /// <exception cref="InvalidOperationException">The <paramref name="dialog"/> is already visible in the window.</exception>
        public static async Task ShowMetroDialogAsync(this MetroWindow window, BaseMetroDialog dialog,
            MetroDialogSettings settings = null)
        {
            window.Dispatcher.VerifyAccess();
            if (window.metroDialogContainer.Children.Contains(dialog))
                throw new InvalidOperationException("The provided dialog is already visible in the specified window.");

            await HandleOverlayOnShow(settings, window);

            dialog.Dispatcher.Invoke(() =>
            {
                SizeChangedEventHandler sizeHandler = SetupAndOpenDialog(window, dialog);
                dialog.SizeChangedHandler = sizeHandler;
            });

            await dialog.WaitForLoadAsync();

            dialog.OnShown();
            if (DialogOpened != null)
            {
                DialogOpened(window, new DialogStateChangedEventArgs());
            }

            await dialog.WaitUnitlClosed();
        }
```

I will create a pull request shortly.

**Comment by karlmtc** [comment]
There's a typo in the method signature:

WaitUnitlClosed()

Should be

WaitUntilClosed()

**Comment by IsNull** [comment]
Yeah it already has been fixed in the pull request #1911
